### PR TITLE
Make alignment compression faster

### LINKED ISSUE #238: Alignment-related operations are slow
And @ognian- has already figured out that it's because of unneeded string copying.

### PR DESCRIPTION
**Pull request by matsen**
Closes #238

### CODE REVIEW

File: src/alignment.hpp [hunk 1]
```diff
@@ -21,7 +21,7 @@ class Alignment {
 
   // Is the alignment non-empty and do all sequences have the same length?
   bool IsValid() const;
-  std::string at(const std::string& taxon) const;
+  const std::string& at(const std::string& taxon) const;
```
Comment (ognian-) [inline src/alignment.hpp:24]: The same change should be made to the definition in alignment.cpp

Comment (matsen) [reply src/alignment.hpp:24]: 🤦 fixed

File: src/site_pattern.hpp [hunk 2]
```diff
@@ -15,8 +15,8 @@
 class SitePattern {
  public:
   SitePattern() = default;
-  explicit SitePattern(const Alignment& alignment, const TagStringMap& tag_taxon_map)
-      : alignment_(alignment), tag_taxon_map_(tag_taxon_map) {
+  SitePattern(const Alignment& alignment, TagStringMap tag_taxon_map)
```
Comment (ognian-) [inline src/site_pattern.hpp:18]: Better change tag_taxon_map to be rvalue (TagStringMap&& tag_taxon_map), since it will be moved from.

Comment (matsen) [reply src/site_pattern.hpp:18]: OK, I'm glad to have you to talk this over. See 

https://stackoverflow.com/questions/51705967/advantages-of-pass-by-value-and-stdmove-over-pass-by-reference

Comment (ognian-) [reply src/site_pattern.hpp:18]: You're right - my comment is not valid. I assumed the constructor is meant to only accept rvalues, but it's used with lvalues in the rest of the code.

**Pull request merged**